File: SudokuSolver_Kotlin/src/main/sudokusolver/kotlin/logic/PruneCandidates.kt
package sudokusolver.kotlin.logic

import sudokusolver.kotlin.Board
import sudokusolver.kotlin.Cell
import sudokusolver.kotlin.RemoveCandidates
import sudokusolver.kotlin.SolvedCell
import sudokusolver.kotlin.SudokuNumber
import sudokusolver.kotlin.UnsolvedCell
import sudokusolver.kotlin.enumIntersect
import java.util.EnumSet

fun pruneCandidates(board: Board<Cell>): List<RemoveCandidates> =
    board.cells.filterIsInstance<UnsolvedCell>().mapNotNull { cell ->
        val sameRow = board.getRow(cell.row)
        val sameColumn = board.getColumn(cell.column)
        val sameBlock = board.getBlock(cell.block)
        (sameRow + sameColumn + sameBlock)
            .filterIsInstance<SolvedCell>()
            .mapTo(EnumSet.noneOf(SudokuNumber::class.java)) { it.value }
            .let { visibleValues -> cell.candidates enumIntersect visibleValues }
            .takeUnless { it.isEmpty() }
            ?.let { toRemove -> RemoveCandidates(cell, toRemove) }
    }

File: SudokuSolver_Swift/SudokuSolver_Swift/logic/simple/PruneCandidates.swift
/*
 * If a cell is solved, then no other cells in the same unit can have that number as a candidate.
 */
func pruneCandidates(board: Board<Cell>) -> [BoardModification] {
    board.cells.unsolvedCells.compactMap { cell in
        let sameRow = board.getRow(rowIndex: cell.row)
        let sameColumn = board.getColumn(columnIndex: cell.column)
        let sameBlock = board.getBlock(blockIndex: cell.block)
        let visibleValues = (sameRow + sameColumn + sameBlock).solvedCells.map(\.value)
        let toRemove = cell.candidates.intersection(visibleValues)
        return toRemove.isEmpty ? nil : BoardModification(cell: cell, candidates: toRemove)
    }
}
